plugins/artifacts/artifacts.go

```diff
@@ -102,6 +173,134 @@ func (tp *taskPlugin) Stopped(result engines.ResultSet) (bool, error) {
 	return !tp.failed.Get(), err
 }
 
+func (tp *taskPlugin) hashArtifact(name string, r io.ReadSeeker) error {
+	// Skip if no COT is to be generated
+	if !tp.createCOT {
+		return nil
+	}
+
+	var err error
+	h := sha256.New()
+	if _, err = io.Copy(h, r); err != nil {
+		return errors.Wrap(err, "failed to hash artifact from reader")
+	}
+	if _, err = r.Seek(0, 0); err != nil {
+		return errors.Wrap(err, "failed to seek artifact reader to start")
+	}
+
+	// Set artifact hash in uploaded for COT generation
+	tp.mUploaded.Lock()
+	defer tp.mUploaded.Unlock()
+	tp.uploaded[name] = h.Sum(nil)
+
+	return nil
+}
+
+func (tp *taskPlugin) Finished(success bool) error {
+	// Skip if no COT is to be generated
+	if !tp.createCOT {
+		return nil
+	}
+
+	// Upload certified log if requested
+	if tp.certifiedLog {
+		const certifiedLogName = "public/logs/certified.log"
+
+		r, err := tp.context.ExtractLog()
+		if err != nil {
+			return err
+		}
+		defer r.Close()
+
+		// Compute artifact hash for chain-of-trust
+		if err = tp.hashArtifact(certifiedLogName, r); err != nil {
+			return err
+		}
+
+		compressed, err := tp.plugin.environment.TemporaryStorage.NewFile()
+		if err != nil {
+			return err
+		}
+		defer compressed.Close()
+
+		// Compress log
+		zipper := gzip.NewWriter(compressed)
+		if _, err = io.Copy(zipper, r); err != nil {
+			return errors.Wrap(err, "failed to compress log")
+		}
+		if err = zipper.Close(); err != nil {
+			return errors.Wrap(err, "failed to close compressing of log")
+		}
+		_, err = compressed.Seek(0, 0)
+		if err != nil {
+			return errors.Wrap(err, "Failed to seek to start of compressed log")
+		}
+
+		// Let's upload from compressed
+		err = tp.context.UploadS3Artifact(runtime.S3Artifact{
+			Name:     certifiedLogName,
+			Mimetype: "text/plain; charset=utf-8",
+			Stream:   compressed,
+			Expires:  tp.context.TaskInfo.Expires,
+			AdditionalHeaders: map[string]string{
+				"Content-Encoding": "gzip",
+			},
+		})
+		if err != nil {
+			err = errors.Wrap(err, "failed to upload certified.log")
+			tp.monitor.Error(err)
+			return runtime.ErrNonFatalInternalError // We don't expect upload errors to be fatal
+		}
+	}
```

Comment:
I am confused for what's the advantage of uploading the log file with a different name, but I guess this predates taskcluster-worker.

Reply:
None... it's just that it's two different plugins, and it's what we do in other workers...

I would love to change artifact names, but it's probably a bit sketchy to do that :)